Initial container support for varats

boehmseb opened the pull request:
Still need to wait for PolyJIT/benchbuild#347 and related issues before merging.

What works:
- [x] declaring/building base containers
- [x] using custom BB/varats versions in the container
- [x] using pip to install BB/varats in the container
- [x] running BB experiments inside a container
- [x] result files are written to specified results directory
- [x]  documentation
- [x] installing research tools into a container
- [x] running containers on slurm (see PolyJIT/benchbuild#322)

boehmseb committed:
container support WIP

boehmseb committed:
add temporary container env setup script until proper config is in place

boehmseb committed:
containers WIP

boehmseb committed:
Add tool for building base containers

boehmseb committed:
Merge branch 'vara-dev' into f-ContainerSupport

boehmseb committed:
containers: JustCompile working

boehmseb committed:
Fix BB install from source getting overridden by varats install

boehmseb committed:
Fix primary source type and remove hack in get_local_project_git_path()

boehmseb committed:
Add todos to upgrade BB version

boehmseb committed:
Merge branch 'vara-dev' into f-ContainerSupport

boehmseb committed:
Docs: fix broken paths and line numbers

boehmseb committed:
Fix typing related docs warnings/errors

boehmseb committed:
Fix some typing problems

boehmseb committed:
Container documentation

boehmseb committed:
Merge branch 'vara-dev' into f-ContainerSupport

boehmseb committed:
Upgrade BB version

boehmseb committed:
Fix pylint warning

boehmseb committed:
Revert "add temporary container env setup script until proper config is in place"

This reverts commit 5fe1d6352f5942179acd4757e3d34f3d0fa934f5.

boehmseb committed:
Fix broken tests

boehmseb committed:
Add issue link to todo

boehmseb committed:
Add method stubs for not yet implemented abstract methods.

Why didn't those tests fail on my machine?

vulder committed:
Merge branch 'vara-dev' into f-ContainerSupport

vulder committed:
Fixes

vulder commented on docs/source/vara-ts-api/containers.rst:
```diff
@@ -0,0 +1,52 @@
+                     +
+    ┌─────────────────────────────────┐
+    │       Experiment Specific       │
+    └─────────────────────────────────┘
```
```suggestion
    ┌───────────────────────────────────┐  \
    │ Base Container (e.g., Debian 10)  │   |
    └───────────────────────────────────┘   |
                      +                     |
    ┌───────────────────────────────────┐   |
    │              varats               │   |
    └───────────────────────────────────┘   |
                      +                      > Base Image
    ┌───────────────────────────────────┐   |
    │            benchbuild             │   |
    └───────────────────────────────────┘   |
                      +                     |
    ┌───────────────────────────────────┐   |
    │    Research Tool (e.g., VaRA)     │   |
    └───────────────────────────────────┘  /
                      +
    ┌───────────────────────────────────┐
    │         Project Specific          │
    └───────────────────────────────────┘
                      +
    ┌───────────────────────────────────┐
    │        Experiment Specific        │
    └───────────────────────────────────┘
```

vulder commented on docs/source/vara-ts-api/projects.rst:
```diff
@@ -182,6 +182,20 @@ also makes it easier to see whether a block can be fixed in the future
+
+    CONTAINER = get_base_image(ImageBase.DEBIAN_10)
+
+To ensure that the container already contains everything required by BenchBuild, you should use one of our base images using the function :func:`~varats.containers.containers.get_base_image()`. You can then add additional layers to the image using BenchBuild's buildah API.
```
Maybe we should add a small example after this how one can add dependencies, e.g., with an apt call.

vulder commented on docs/source/vara-ts-api/tables.rst:
```diff
@@ -25,7 +25,7 @@ Table helper modules
 .............
 
-.. automodule:: varats.tables.table
+.. automodule:: varats.table.table
```
Is this the adaption the make the folder naming schema consistent? How is this related to the container PR?

boehmseb replied:
It is an oversight from that PR that I fixed while I was at it.

vulder commented on docs/source/vara-ts-api/tools/vara-container.rst:
```diff
@@ -0,0 +1,14 @@
+.. program-output:: vara-container -h
+    :nostderr:
+
+TODO: add section about tool selection once implemented
```
Do you want to fix this before merging? Otherwise, please slap an issue onto the TODO :)

vulder commented on docs/source/vara-ts/benchbuild.rst:
```diff
@@ -87,3 +87,49 @@ To adapt and tune BenchBuild further, you can moify the different configuration
+BenchBuild can run its experiments inside a container.
+This allows to customize the execution environment on a per-project(-version) and per-experiment level.
+
+To use BenchBuild's container support, you first need to setup `buildah <https://github.com/containers/buildah/blob/master/install.md>`_ and `podman <https://podman.io/getting-started/installation>`_ on your system.
```
```suggestion
To use BenchBuild's container support, you first need to setup `buildah <https://github.com/containers/buildah/blob/master/install.md>`_ and `podman <https://podman.io/getting-started/installation>`_ on your system.
Please follow their install instructions on how to setup both tools.
Keep in mind that you have to set a subuid and subgid mapping on all machines that need to run containers. 
```

vulder commented on docs/source/vara-ts/benchbuild.rst:
```diff
@@ -87,3 +87,49 @@ To adapt and tune BenchBuild further, you can moify the different configuration
+To use BenchBuild's container support, you first need to setup `buildah <https://github.com/containers/buildah/blob/master/install.md>`_ and `podman <https://podman.io/getting-started/installation>`_ on your system.
+
+Then, you have to set the following parameters in the :ref:`BenchBuild config <How-to configure BenchBuild yourself>`:
+
```
We wanted to reword/rework this so these options are only presented to the reader as config knobs. The current settings are now generated by default.

vulder commented on varats-core/varats/utils/settings.py:
```diff
@@ -33,6 +33,25 @@
+    "from_source": {
+        "desc":
+            "Whether to install varats in the container from a local "
+            "checkout or pip.",
```
```suggestion
            "source checkout or pip.",
```

vulder commented on varats-core/varats/utils/settings.py:
```diff
@@ -33,6 +33,25 @@
+        "default": False
+    },
+    "varats_source": {
+        "desc": "The local checkout of varats to use for the source install.",
```
```suggestion
        "desc": "Path to the local checkout of varats to use for the source install.",
```

vulder commented on varats/varats/containers/containers.py:
```diff
@@ -0,0 +1,159 @@
+        )
+}
+
+VARATS_ROOT = "/varats_root/"
+BB_ROOT = "/app/"
```
Maybe prefix this with `CONTAINER_`, otherwise, this could be confused with the root outside of the container.

vulder replied:
Further, it could be helpful to declare these vars as a `Path`, because they represent paths and the following code makes a lot of modifications, where when some `/` is not correctly added the resulting string will result in a crash.

vulder commented:
List of other things we found yesterday:

- [ ] add bulidah/podman aliases to docs + small helper section how to manually access containers
- [x] container registry clear (will be handled by BB)
- [ ] add crun dependency to docs
- [x] research_tool selection (already in TODOs)

boehmseb committed:
Apply suggestions from code review

Co-authored-by: Florian Sattler <vuld3r@gmail.com>

boehmseb committed:
Make vara-container more useful

boehmseb committed:
Add small container building example.

boehmseb committed:
Implement feedback about documentation

boehmseb committed:
containers.py: use paths instead of strings

boehmseb committed:
Add missing docstrings

boehmseb committed:
Improve BB container documentation

boehmseb committed:
Fix tests and run updated pre-commit

boehmseb commented:
> List of other things we found yesterday:
>     * [ ]  small helper section how to manually access containers

You still know how you did this right? Because I did never really do this.
 
>     * [ ]  BB config not correctly found if not duplicated in vara-root

I cannot reproduce this.

boehmseb committed:
Fix pylint errors

boehmseb committed:
Add image_name property to ImageBase

boehmseb committed:
Allow to export base containers to the file system

vulder committed:
Add deps to container for xz

vulder committed:
Remove unneeded PyQt cap

vulder committed:
Store BB_JOBS config to circumvent BB thread detection bug

boehmseb committed:
Add container debug how-to

vulder committed:
Config warning

Sinerum committed:
fix on xz compile

Sinerum committed:
test xz make change

boehmseb committed:
Update to new BB message bus

Sinerum committed:
revert changes to xz project file

boehmseb committed:
Merge branch 'vara-dev' into f-ContainerSupport

boehmseb committed:
Merge branch 'vara-dev' into f-ContainerSupport

boehmseb committed:
Enable vara container install (#441)

resolves se-passau/VaRA#718
The dependency declarations in this PR will also be useful for se-passau/VaRA#626

boehmseb commented:
CI will work again when BB releases a new version.

vulder committed:
Fixes tool name look up on the slurm side

boehmseb committed:
Merge branch 'vara-dev' into f-ContainerSupport

boehmseb committed:
Update BB container docs

boehmseb committed:
Use BBs delete command to delete base images

boehmseb committed:
Allow interactive debugging when running vara-container build

boehmseb committed:
BB container docs: fix indendation

vulder committed:
Merge branch 'vara-dev' into f-ContainerSupport

boehmseb committed:
Fix crash when container/from_source is false and container/source_dir is not set.

boehmseb committed:
Add docs on how to use containers with slurm (#446)

boehmseb committed:
Fix typo in docs

boehmseb committed:
Fix LD_LIBRARY_PATH in container

boehmseb committed:
Upgrade BenchBuild requirement to 6.1.1

boehmseb committed:
Merge branch 'vara-dev' into f-ContainerSupport

boehmseb committed:
Upgrade click version to 8.0.0

boehmseb committed:
Add default values to `vara-container prepare-slurm`

boehmseb committed:
Add container `dev_mode` option

boehmseb committed:
Add missing changes to slurm script template

boehmseb committed:
Fix type issues

boehmseb committed:
Fix pylint warnings

boehmseb committed:
Overwrite exported images

boehmseb committed:
Add missing jinja2 requirement

boehmseb committed:
Remove jinja2 version requirement for now

boehmseb committed:
Add missing docstring

Sinerum committed:
Fix type of export_path, needs to be Path was string

boehmseb committed:
Merge branch 'vara-dev' into f-ContainerSupport

boehmseb committed:
Remove unnecessary code

boehmseb committed:
Test container stuff

boehmseb committed:
Merge branch 'vara-dev' into f-ContainerSupport

boehmseb committed:
Set Jinja version to >=3.0.1

boehmseb committed:
Reformat too long line

boehmseb committed:
Fix duplicate test name and add docs to tests

vulder commented:
resolves se-sic/VaRA#718

boehmseb committed:
Silence pylint

vulder commented on varats/setup.py:
```diff
@@ -22,9 +22,10 @@
         "argparse-utils>=1.2.0",
-        "benchbuild>=6.0.1",
-        "click>=7.0.0",
+        "benchbuild>=6.1.1",
```
```suggestion
        "benchbuild>=6.2.0",
```
Following the requirements.txt this should be 6.2

boehmseb committed:
Remove issue link in phasar container install function

boehmseb committed:
Update varats/setup.py

Co-authored-by: Florian Sattler <vuld3r@gmail.com>
